### PR TITLE
#202 Implementing Range#bsearch API

### PR DESCRIPTION
**Pull request by Maxwell-Alexius**
This PR aims to implement binary search method in Range API from Ruby 2.4.1 to Goby Lang

### CODE REVIEW

File: vm/range.go [hunk 1]
```diff
@@ -306,4 +306,109 @@ var builtInRangeInstanceMethods = []*BuiltInMethodObject{
...
+							return NULL
+						}
+
+						if value > 0 {
+							start = mid + 1
+						} else {
+							end = mid - 1
+						}
+					default:
+						return newError("Must be integer or boolean type. got=%T", result)
```
Comment (st0012) [inline vm/range.go]: Please use this method instead https://github.com/goby-lang/goby/blob/master/vm/error.go#L59

File: vm/range.go [hunk 2]
```diff
@@ -306,4 +306,109 @@ var builtInRangeInstanceMethods = []*BuiltInMethodObject{
...
+							return initIntegerObject(pivot)
+						}
+
+						if value {
+							end = mid - 1
+						} else {
+							start = mid + 1
+						}
+					case *IntegerObject:
+						fmt.Println(start + end)
```
Comment (st0012) [inline vm/range.go]: Remember to remove this line

File: vm/range.go [hunk 3]
```diff
@@ -306,4 +306,109 @@ var builtInRangeInstanceMethods = []*BuiltInMethodObject{
...
+				start := ran.Start
+				end := ran.End
+				var mid int
+				pivot := -1
+
+				for {
+					if (start+end)%2 != 0 {
+						mid = (start+end)/2 + 1
+					} else {
+						mid = (start + end) / 2
```
Comment (st0012) [inline vm/range.go]: Should there be spaces around `+`? Make it consistent like above lines.

File: vm/range.go [hunk 4]
```diff
@@ -306,4 +306,109 @@ var builtInRangeInstanceMethods = []*BuiltInMethodObject{
...
+					}
+					result := t.builtInMethodYield(blockFrame, initIntegerObject(mid))
+
+					switch result.Target.(type) {
```
Comment (st0012) [inline vm/range.go]: You can write `r := result.Target.(type)` then `r` will be assert to the matched type.

File: vm/range.go [hunk 5]
```diff
@@ -306,4 +306,109 @@ var builtInRangeInstanceMethods = []*BuiltInMethodObject{
...
+					case *BooleanObject:
+						value := result.Target.(*BooleanObject).Value
```
Comment (st0012) [inline vm/range.go]: So here you can just `value := r.Value`, Go already assumes `r` is `*BooleanObject` type.

File: vm/range.go (hunk 3 above)
Comment (Maxwell-Alexius) [inline vm/range.go]: I found out that when executing:

```
$ go test ./...
```

It just auto-corrected the second `mid` assignment which makes the space wraps around the `+` sign

File: vm/range.go (hunk 1 above)
Comment (Maxwell-Alexius) [inline vm/range.go]: Can I extend something like `WrongTypeErrorClass`?

Comment (Maxwell-Alexius) [inline vm/range.go]: Oops, I saw `TypeError` implemented, sorry XD

File: vm/range.go (hunk 3 above)
Comment (st0012) [inline vm/range.go]: You need to type -s -w

File: vm/range.go [hunk 6]
```diff
@@ -306,4 +304,106 @@ var builtInRangeInstanceMethods = []*BuiltInMethodObject{
...
+							return NULL
+						}
+
+						if r.Value > 0 {
+							start = mid + 1
+						} else {
+							end = mid - 1
+						}
+					default:
+						return initErrorObject(TypeErrorClass, "Expect Integer or Boolean type. got=%v", r)
```
Comment (Maxwell-Alexius) [inline vm/range.go]: @st0012 Can you help me check it. I'm not quite sure this error implementation is right or wrong.

Comment (st0012) [inline vm/range.go]: Yes, you can open REPL and make this method fail, if the REPL doesn't crash and you see the error message means it works.

Comment (Maxwell-Alexius) [inline vm/range.go]: Thanks, let me try it

**Comment by st0012** [comment]
LGTM, @hachi8833 do you want to review this?

**Pull request merged**